src/libsystemd/sd-json/json-util.c

```diff
@@ -144,3 +146,133 @@ int json_dispatch_path(const char *name, sd_json_variant *variant, sd_json_dispa
+                return r;
+
+        if (pidref->fd_id > 0) {
+                /* If we have the pidfd inode number, also acquire the boot ID, to make things universally unique */
+                r = sd_id128_get_boot(&boot_id);
+                if (r < 0)
+                        return r;
+        }
+
+        return sd_json_buildo(
+                        ret,
+                        SD_JSON_BUILD_PAIR_INTEGER("pid", pidref->pid),
+                        SD_JSON_BUILD_PAIR_CONDITION(pidref->fd_id > 0, "pidfdId", SD_JSON_BUILD_INTEGER(pidref->fd_id)),
+                        SD_JSON_BUILD_PAIR_CONDITION(pidref->fd_id > 0, "bootId", SD_JSON_BUILD_ID128(boot_id)));
+}
+
+int json_dispatch_pidref(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        PidRef *p = ASSERT_PTR(userdata);
+        int r;
+
+        assert(variant);
+
+        /* Turns a JSON PID triplet back into a PidRef, i.e. the reverse of json_variant_new_pidref()
+         * above. If SD_JSON_STRICT is set will acquire a pidfd for the process, and validate that the
+         * auxiliary fields match it. Otherwise, will only do a */
+
+        if (sd_json_variant_is_null(variant)) {
+                pidref_done(p);
+                return 0;
+        }
+
+        struct {
+                uint64_t pid, fd_id;
+                sd_id128_t boot_id;
+        } data = {};
+
+        if (sd_json_variant_is_integer(variant))
+                /* Permit a simple classic integer based format */
+                data.pid = sd_json_variant_integer(variant);
+        else if (sd_json_variant_is_string(variant)) {
+                /* As usual, allow integers be encoded as strings too */
+                r = safe_atou64(sd_json_variant_string(variant), &data.pid);
+                if (r < 0)
+                        return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not numeric PID", strna(name));
+        } else if (sd_json_variant_is_object(variant)) {
+
+                static const sd_json_dispatch_field dispatch_table[] = {
+                        { "pid",     _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64, voffsetof(data, pid),     SD_JSON_MANDATORY },
+                        { "pidfdId", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64, voffsetof(data, fd_id),   0                 },
+                        { "bootId",  SD_JSON_VARIANT_STRING,        sd_json_dispatch_id128,  voffsetof(data, boot_id), 0                 },
+                        {}
+                };
+
+                r = sd_json_dispatch(variant, dispatch_table, flags, &data);
+                if (r < 0)
+                        return r;
+        } else
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not numeric PID or a PID structure.", strna(name));
+
+        if (data.pid > PID_T_MAX || !pid_is_valid(data.pid))
```

Comment:
Hmm, the PID_T_MAX check can probably be folded into pid_is_valid?

Reply:
data.pid is uint64_t, hence we need to check with PID_T_MAX at first.

Reply:
it is there already, but it's not redundant here because pid_t is 32bit, and data.pid is 64bit, hence before we cast this down we have to check it's in the 32bit range. will add comment